mixmatch/proxy.py

```diff
@@ -53,60 +54,74 @@ def __init__(self, method, path, headers):
-                self.action.insert(len(self.action), 'detail')
-        else:
-            raise ValueError
 
-        if self.method in ['GET']:
-            self.stream = True
-        else:
-            self.stream = False
+        # if we receive a request with no action
+        # don't forward it, but return our
+        # list of supported api versionn
+        if len(self.request_path) > 2:
+            self.version = self.request_path[1]
+
+            self.detailed = True
+            if self.service_type == 'image':
+                # /image/{version}/{action}
+                self.action = self.request_path[2:]
+            elif self.service_type == 'volume':
+                # /volume/{version}/{project_id}/{action}
+                self.action = self.request_path[3:]
+
+                # if request is to /volumes, change it
+                # to /volumes/detail for aggregation
+                if self.method == 'GET' \
+                        and self.action[-1] == 'volumes':
+                    self.detailed = False
+                    self.action.insert(len(self.action), 'detail')
+            else:
+                raise ValueError
 
-        resource_id = None
-        mapping = None
-        aggregate = False
+            if self.method in ['GET']:
+                self.stream = True
+            else:
+                self.stream = False
+
+            resource_id = None
+            mapping = None
+            aggregate = False
+
+            if len(self.action) > 1 and is_valid_uuid(self.action[1]):
+                resource_id = self.action[1]
+                mapping = model.ResourceMapping.find(
+                    resource_type=self.action[0],
+                    resource_id=resource_id)
+            else:
+                if self.method == 'GET' \
+                   and self.action[0] in ['images', 'volumes', 'snapshots']:
+                    aggregate = True
+
+            self.local_token = headers['X-AUTH-TOKEN']
+            LOG.info('Local Token: %s ' % self.local_token)
+
+            if 'MM-SERVICE-PROVIDER' in headers and 'MM-PROJECT-ID' in headers:
+                # The user wants a specific service provider, use that SP.
+                self.service_provider = headers['MM-SERVICE-PROVIDER']
+                self.project_id = headers['MM-PROJECT-ID']
+                self._forward = self._targeted_forward
+            elif aggregate:
+                self._forward = self._aggregate_forward
+            elif mapping:
+                # Which we already know the location of, use that SP.
+                self.service_provider = mapping.resource_sp
+                self.project_id = mapping.tenant_id
+                self._forward = self._targeted_forward
+            else:
+                self._forward = self._search_forward
+
+        elif len(self.request_path) < 2:
+            # calls with no version and action
+            self._forward = self._list_api_versions
```

Comment:
Can you put this part first?  That way, you can put a return in the ``if`` block and avoid indenting everything that comes after it